### PR TITLE
Fix reboot

### PR DESCRIPTION
**Pull request by haslinghuis**
- [x] Change `reinitializeConnection` to use `setTimeout` as `GUI.timeout_add` does not work here.
- [x] `reinitializeConnection` should be called in the callback of the MSP send command.
- [x] Move clicks variable from DOM to JS.
- [x] Don't send MSP command if there is no connection ID.
- [x] Close connection graceful in `reinitializeConnection` if it's still available. This is an API recommendation.
- [x] Added different timeout for MacOS (Thanks to @ctzsnooze)
- [x] Fixed some logging (@ctzsnooze)

### CODE REVIEW

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> FAIL
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> FAIL
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> FAIL
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Comment by ctzsnooze** [comment]
What improvement should I expect to see from this?

**Comment by haslinghuis** [comment]
@ctzsnooze while working on an issue I found out that console.log did not produce any output if used with current code. However it did not solve the issue and there are no other reports.

**Comment by ctzsnooze** [comment]
@haslinghuis is this PR still active and still required?

**Comment by haslinghuis** [comment]
Not yet, as if it ain't broke - don't fix it.

Testing this function is looks it's not executing (a console.log does not occur, changes in the PR does allow code to execute).
Ordering some hardware for more testing soon. This issue puzzles me: https://github.com/betaflight/betaflight-configurator/issues/2796

**Comment by DHaacke** [comment]
@haslinghuis I tried your changes (replacing the GUI.timeout with setTimeout and changing the connectionTimeout to 1500) and there was no difference. I would add to my notes that the FC seems to initialize and I get the normal startup beeps, but instead of connecting almost immediately, the correct port appears (briefly) as though its connected, but is replaced with the DFU LOADER after a good 3 seconds or so. Hope that helps, and very much appreciate your efforts!

**Comment by haslinghuis** [comment]
Thanks. At least we have a workaround powering up the board before connecting. Have no clue yet as the race between USB and GPS for serial connection should not behave like this. My SPRacing H7 Extreme with GPS just works.

**Comment by DHaacke** [comment]
Thank YOU!  If you’d like me try anything, I’m a developer by trade myself
(Typescript, Javascript and Java mostly) but not a Betaflight dev
obviously. Just recently cloned the repo and started poking around. Anyway,
I’d be happy to try anything/anytime. Just holler.  Also, I do’t recall if
I mentioned in my notes, but I do work on Mac.

Sure appreciate all you do for this community!

-Doug

**Comment by haslinghuis** [comment]
Mac is the one machine I don't have. As I have updated my development environment to 11th gen I5 with 32GB RAM and 1TB samsung 980 SSD using Linux.

Mac has disappointed me in odd behavior with clang differentiation (firmware) and GUI issues with hide / minimize

**Comment by DHaacke** [comment]
I hear you.  I was hard-core Microsoft C/C++ dev for many years coming
from DOS, then Windows and through the Visual Studio beginnings. I sold my
business to a friendly competitor, who has contracted me for the last 10 or
12 years. They had me write an iOS app for them, and I fell in love with
the Mac.  Does everything I need.  I will admit to using VS Code, though,
which I’’m finding to be pretty sweet. As I get older, I’m finding the
minimalist code and IDEs more appealing.

-Doug

**Comment by haslinghuis** [comment]
Curious if you could check the issue is OSX only

**Comment by DHaacke** [comment]
Will do, sir.

**Comment by haslinghuis** [comment]
Builds here: https://dev.azure.com/Betaflight/Betaflight%20Nightlies/_build/results?buildId=5902&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by DHaacke** [comment]
Thanks for the link to the nightly!   Unfortunately, the nightly showed
the same behavior. It briefly shows a connect with the correct port, then
jumps right to DFU, all within a second.

-Doug

**Comment by haslinghuis** [comment]
With that background you could take a look at debugging it yourself as I don't have OSX running.
I have updated lots of documentation to get started.

Please try disabling auto-connect and also try a Windows or Linux computer to rule out the board.

Usual it's the other way around, people can't get into DFU mode.

**Comment by DHaacke** [comment]
I will do that. I’ve been trying to familiarize myself with the code. It
looks pretty straightforward.

Where can I find the documentation you’ve updated?  Is it in the WIKI or
elsewhere?

Thanks for all your help and advice, as always. I’ll keep you posted with
my findings...

-Doug

**Comment by haslinghuis** [comment]
https://github.com/betaflight/betaflight/blob/master/docs/development/Building%20in%20Mac%20OS%20X.md
https://github.com/betaflight/betaflight/wiki/Betaflight-Contribution

**Comment by haslinghuis** [comment]
Updated artifacts can be found here: https://dev.azure.com/Betaflight/Betaflight%20Nightlies/_build/results?buildId=5914&view=artifacts&pathAsName=false&type=publishedArtifacts

File: src/js/serial_backend.js
Comment (McGiverGim) [inline src/js/serial_backend.js:47]: Another fix for this can be:
```javascript
const clicks = thisElement.data('clicks') || false;
```
but I think is better to use variables and not attach variables to html elements.

**Comment by McGiverGim** [comment]
This will go in 10.8? I don't see a tag.
If true, I think we need to merge this and the update to NW 0.60.0 to start with something "real" to look at the other bugs before the final version.

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> FAIL
 - approver count at least three -> FAIL

**Review by blckmn — APPROVED** [review]

**Comment by ctzsnooze** [comment]
This seems OK on my Mac, however... there have been longstanding issues with flashing a local hex. With this PR:

I can connect to any board easily, and disconnect, without console errors.

There are issues with flashing local files.  Often after clicking flash, we get into DFU mode but stop there.  I have to click flash a second time for the flash to start.

This is a transcript of a flash that worked:

```
serial: connection opened with ID: 21 , Baud: 115200
serial_backend.js:248 Requesting configuration data
MSPHelper.js:1687 Arming disable
MSPHelper.js:1690 Real time clock set
three.min.js:179 THREE.WebGLRenderer 97
serial.js:455 serial: device_lost ID: 21 (receive)
errorHandler @ serial.js:455
watch_for_on_receive_errors @ serial.js:114
serial.js:249 serial: closed connection with ID: 21, Sent: 1911 bytes, Received: 20875 bytes
port_handler.js:129 PortHandler - Removed: [{"path":"/dev/cu.usbmodem0x80000001","displayName":"Betaflight STM32F411","vendorId":1155,"productId":22336},{"path":"/dev/tty.usbmodem0x80000001","displayName":"Betaflight STM32F411","vendorId":1155,"productId":22336}]
FirmwareCache.js:234 Firmware cache journal loaded; number of entries: 0
firmware_flasher-e75c1fe2.js:247 loaded some builds for later
firmware_flasher-e75c1fe2.js:269 unified config cached data 11 mins old
firmware_flasher-e75c1fe2.js:567 board changed to FLYWOOF411
firmware_flasher-e75c1fe2.js:672 We have the config cached for FLYWOOF411+FLWO
firmware_flasher-e75c1fe2.js:1001 Loading file from: ~/Documents/Personal/RC/betaflight_ctzsnooze/obj/betaflight_4.3.0_STM32F411_1853a25c0.hex
firmware_flasher-e75c1fe2.js:1009 File loaded (1395281)
hex_parser.js:104 HEX_PARSER - File parsed in: 75.494140625 ms
ConfigInserter.js:104 Custom defaults inserted in: 0.6279296875 ms
stm32usbdfu.js:98 USB DFU detected with ID: 12
stm32usbdfu.js:124 Device opened with Handle ID: 6
stm32usbdfu.js:157 Claimed interface: 0
common.js:1537 Chrome API Error: Transfer stalled..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:1537:90)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:300:13
checkChromeRuntimeError @ common.js:1537
(anonymous) @ stm32usbdfu.js:300
stm32usbdfu.js:301 USB getFunctionalDescriptor failed! 4
stm32usbdfu.js:646 Using transfer size: 2048
stm32usbdfu.js:781 Address load for option bytes sector succeeded.
stm32usbdfu.js:726 Option bytes read successfully
stm32usbdfu.js:727 Chip does not appear read protected
stm32usbdfu.js:833 Executing local chip erase (8) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]
stm32usbdfu.js:856 Erasing. sector 0, page 0 @ 0x8000000
stm32usbdfu.js:856 Erasing. sector 0, page 1 @ 0x8004000
stm32usbdfu.js:856 Erasing. sector 0, page 2 @ 0x8008000
stm32usbdfu.js:856 Erasing. sector 0, page 3 @ 0x800c000
stm32usbdfu.js:856 Erasing. sector 1, page 0 @ 0x8010000
stm32usbdfu.js:856 Erasing. sector 2, page 0 @ 0x8020000
stm32usbdfu.js:856 Erasing. sector 2, page 1 @ 0x8040000
stm32usbdfu.js:856 Erasing. sector 2, page 2 @ 0x8060000
stm32usbdfu.js:843 Erase: complete
stm32usbdfu.js:913 Writing data ...
stm32usbdfu.js:971 Writing: done
stm32usbdfu.js:985 Verifying data ...
stm32usbdfu.js:550 Verification successful, matching: 8788 bytes
stm32usbdfu.js:550 Verification successful, matching: 8 bytes
stm32usbdfu.js:550 Verification successful, matching: 2601 bytes
stm32usbdfu.js:550 Verification successful, matching: 32768 bytes
6stm32usbdfu.js:550 Verification successful, matching: 65536 bytes
stm32usbdfu.js:550 Verification successful, matching: 61243 bytes
stm32usbdfu.js:1052 Programming: SUCCESSFUL
stm32usbdfu.js:1107 Script finished after: 40.785 seconds
stm32usbdfu.js:171 Released interface: 0
common.js:1537 Chrome API Error: No such connection..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:1537:90)
    at closed (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:133:13)
checkChromeRuntimeError @ common.js:1537
closed @ stm32usbdfu.js:133
stm32usbdfu.js:134 Failed to close USB device!
stm32usbdfu.js:139 Device closed with Handle ID: 6
port_handler.js:169 PortHandler - Found: [{"path":"/dev/cu.usbmodem0x80000001","displayName":"Betaflight STM32F411","vendorId":1155,"productId":22336},{"path":"/dev/tty.usbmodem0x80000001","displayName":"Betaflight STM32F411","vendorId":1155,"productId":22336}]
port_handler.js:283 Porthandler detected device Betaflight STM32F411 on port: /dev/tty.usbmodem0x80000001
serial.js:119 serial: connection opened with ID: 22 , Baud: 115200
firmware_flasher-e75c1fe2.js:824 Requesting board information
firmware_flasher-e75c1fe2.js:829 1.44.0
```

**Comment by ctzsnooze** [comment]
This is a console log of what happens after clicking 'load firmware' and then 'flash'... up to the point where the flashing stops with the board in DFU mode:

```
Connecting to: /dev/tty.usbmodem0x80000001
serial.js:119 serial: connection opened with ID: 2 , Baud: 115200
serial_backend.js:248 Requesting configuration data
MSPHelper.js:1687 Arming disable
MSPHelper.js:1690 Real time clock set
three.min.js:179 THREE.WebGLRenderer 97
serial.js:249 serial: closed connection with ID: 2, Sent: 21244 bytes, Received: 259138 bytes
2FirmwareCache.js:234 Firmware cache journal loaded; number of entries: 0
firmware_flasher-e75c1fe2.js:247 loaded some builds for later
firmware_flasher-e75c1fe2.js:269 unified config cached data 18 mins old
firmware_flasher-e75c1fe2.js:567 board changed to FLYWOOF411
firmware_flasher-e75c1fe2.js:672 We have the config cached for FLYWOOF411+FLWO
firmware_flasher-e75c1fe2.js:1001 Loading file from: ~/Documents/Personal/RC/betaflight_ctzsnooze/obj/betaflight_4.3.0_STM32F411_1853a25c0.hex
firmware_flasher-e75c1fe2.js:1009 File loaded (1395281)
ConfigInserter.js:104 Custom defaults inserted in: 0.595947265625 ms
serial.js:119 serial: connection opened with ID: 3 , Baud: 115200
MSPConnector.js:54 Connected
stm32.js:158 Looking for capabilities via MSP
stm32.js:168 no flash bootloader detected
MSPConnector.js:70 Disconnected
serial.js:249 serial: closed connection with ID: 3, Sent: 19 bytes, Received: 91 bytes
port_handler.js:129 PortHandler - Removed: [{"path":"/dev/cu.usbmodem0x80000001","displayName":"Betaflight STM32F411","vendorId":1155,"productId":22336},{"path":"/dev/tty.usbmodem0x80000001","displayName":"Betaflight STM32F411","vendorId":1155,"productId":22336}]
serial.js:145 serial: failed to open serial port
```
here is the transcript after I click 'flash firmware' a second time:
```
USB DFU detected with ID: 1
stm32usbdfu.js:124 Device opened with Handle ID: 1
stm32usbdfu.js:157 Claimed interface: 0
common.js:1537 Chrome API Error: Transfer stalled..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:1537:90)
    at chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:300:13
checkChromeRuntimeError @ common.js:1537
(anonymous) @ stm32usbdfu.js:300
stm32usbdfu.js:301 USB getFunctionalDescriptor failed! 4
stm32usbdfu.js:646 Using transfer size: 2048
stm32usbdfu.js:781 Address load for option bytes sector succeeded.
stm32usbdfu.js:726 Option bytes read successfully
stm32usbdfu.js:727 Chip does not appear read protected
stm32usbdfu.js:833 Executing local chip erase (8) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]
stm32usbdfu.js:856 Erasing. sector 0, page 0 @ 0x8000000
stm32usbdfu.js:856 Erasing. sector 0, page 1 @ 0x8004000
stm32usbdfu.js:856 Erasing. sector 0, page 2 @ 0x8008000
stm32usbdfu.js:856 Erasing. sector 0, page 3 @ 0x800c000
stm32usbdfu.js:856 Erasing. sector 1, page 0 @ 0x8010000
stm32usbdfu.js:856 Erasing. sector 2, page 0 @ 0x8020000
stm32usbdfu.js:856 Erasing. sector 2, page 1 @ 0x8040000
stm32usbdfu.js:856 Erasing. sector 2, page 2 @ 0x8060000
stm32usbdfu.js:843 Erase: complete
stm32usbdfu.js:913 Writing data ...
```
then the flash is successful and I remain at the flashing tab with the following transcript:
```
Writing: done
stm32usbdfu.js:985 Verifying data ...
stm32usbdfu.js:550 Verification successful, matching: 8788 bytes
stm32usbdfu.js:550 Verification successful, matching: 8 bytes
stm32usbdfu.js:550 Verification successful, matching: 2601 bytes
stm32usbdfu.js:550 Verification successful, matching: 32768 bytes
6stm32usbdfu.js:550 Verification successful, matching: 65536 bytes
stm32usbdfu.js:550 Verification successful, matching: 61243 bytes
stm32usbdfu.js:1052 Programming: SUCCESSFUL
stm32usbdfu.js:1107 Script finished after: 40.191 seconds
stm32usbdfu.js:171 Released interface: 0
common.js:1537 Chrome API Error: No such connection..
 Traced Error
    at checkChromeRuntimeError (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/utils/common.js:1537:90)
    at closed (chrome-extension://ajdoecjhplhoahggfmgflfnlechcljoc/js/protocols/stm32usbdfu.js:133:13)
checkChromeRuntimeError @ common.js:1537
closed @ stm32usbdfu.js:133
stm32usbdfu.js:134 Failed to close USB device!
stm32usbdfu.js:139 Device closed with Handle ID: 1
port_handler.js:169 PortHandler - Found: [{"path":"/dev/cu.usbmodem0x80000001","displayName":"Betaflight STM32F411","vendorId":1155,"productId":22336},{"path":"/dev/tty.usbmodem0x80000001","displayName":"Betaflight STM32F411","vendorId":1155,"productId":22336}]
port_handler.js:283 Porthandler detected device Betaflight STM32F411 on port: /dev/tty.usbmodem0x80000001
serial.js:119 serial: connection opened with ID: 4 , Baud: 115200
firmware_flasher-e75c1fe2.js:824 Requesting board information
firmware_flasher-e75c1fe2.js:829 1.44.0
serial.js:249 serial: closed connection with ID: 4, Sent: 12 bytes, Received: 77 bytes
```
At this point, the correct betaflight serial port is shown in the serial dropdown as `/dev/tty.usbmodem0x80000001`.  I don't know if it is supposed to open the connection, then close it.

So there are definitely some errors happening here.  It would be awesome if flashing could work better on a mac.

**Comment by ctzsnooze** [comment]
All that stuff is fairly 'normal' I think for flashing on a mac

**Comment by haslinghuis** [comment]
@ctzsnooze this PR is not about flashing. For that we should consider https://github.com/betaflight/betaflight/issues/11334

**Review by blckmn — DISMISSED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> FAIL
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> FAIL
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Review by McGiverGim — APPROVED** [review]
I've not tested it, but seems good to me...

**Comment by haslinghuis** [comment]
@McGiverGim - I need to remove the heavy logging before merging. Will the previous build be available for the user to test?

**Comment by McGiverGim** [comment]
I think so, but I'm not usually sure 😃

**Comment by haslinghuis** [comment]
Removed the logging. Let's see if the other builds will be still there ;-)

**Comment by haslinghuis** [comment]
Removed logging, ready to merge

File: src/js/serial_backend.js
Comment (limonspb) [inline src/js/serial_backend.js:856]: callback?.() :)

Comment (haslinghuis) [reply src/js/serial_backend.js]: Fixed :)

Comment (haslinghuis) [reply src/js/serial_backend.js]: Rebased on master

**Review by McGiverGim — DISMISSED** [review]

**Comment by Asizon** [comment]
Fixes https://github.com/betaflight/betaflight-configurator/issues/2731

**Review by limonspb — DISMISSED** [review]

File: src/js/port_handler.js
Comment (Asizon) [inline src/js/port_handler.js:26]: Seems that now we cant show virtual mode connection available.

**Review by Asizon — CHANGES_REQUESTED** [review]

File: src/js/port_handler.js
Comment (haslinghuis) [reply src/js/port_handler.js:26]: You have to enable Virtual Mode in options tab and restart configurator.

Comment (Asizon) [reply src/js/port_handler.js:26]: I tried that with no succes @haslinghuis

Comment (haslinghuis) [reply src/js/port_handler.js]: Thanks for testing. Reverted the change in Porthandler and now it works again. I'd moved the two lines out of the check because it's being called twice a second to reduce workload.

**Review by limonspb — DISMISSED** [review]

**Comment by ctzsnooze** [comment]
Seems to make Mac flashing more reliable.
MacOS users need a longer timeout in protocols/stm32.js line 121 to get a stable DFU connection before flashing, ie change to:
```
setTimeout(startFlashing, 5000);
```

**Comment by haslinghuis** [comment]
Thanks @ctzsnooze - have added the timeout for MacOS only,

**Review by Asizon — APPROVED** [review]

**Review by ctzsnooze — APPROVED** [review]

**Review by McGiverGim — APPROVED** [review]

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by blckmn** [comment]
AUTOMERGE: (FAIL)
 - github identifies PR as mergeable -> FAIL
 - assigned to a milestone -> PASS
 - cooling off period lapsed -> PASS
 - commit count less or equal to three -> PASS
 - `Don't merge` label NOT found -> PASS
 - at least one `RN:` label found -> PASS
 - `Tested` label found -> PASS
 - assigned to an approver -> PASS
 - approver count at least three -> FAIL

**Comment by McGiverGim** [comment]
Merged. We need to start testing this.

**Pull request merged**
